File: android/src/main/java/io/github/malopezr7/bridgekit/runtime/BridgeValue.kt
package io.github.malopezr7.bridgekit.runtime

/**
 * Availability wrapper for bridgekit state values and consume proxies.
 *
 * - [Available] — a live provider is present; value comes from the provider.
 * - [Initial] — no provider yet (or not yet connected); value is the DSL-declared initial.
 * - [Replacing] — provider is reconnecting; last known value is still accessible but marked
 *   stale. Transitions to [Available] if re-provision arrives, or to [Unprovided] after the
 *   grace window expires. Consumers should treat this as "stale" (W2-3).
 * - [Unprovided] — provider existed but was closed/disconnected; [lastKnown] carries the
 *   last observed value for graceful degradation.
 */
sealed class BridgeValue<out T> {
    data class Available<out T>(val value: T) : BridgeValue<T>()
    data class Initial<out T>(val value: T) : BridgeValue<T>()
    /**
     * Stale-but-accessible state during the replacing grace window.
     * [lastKnown] is the last value from the closing provider.
     */
    data class Replacing<out T>(val lastKnown: T?) : BridgeValue<T>()
    data class Unprovided<out T>(val lastKnown: T?) : BridgeValue<T>()

    /** Unwrap the value regardless of status, returning null for Unprovided with no lastKnown. */
    fun valueOrNull(): T? = when (this) {
        is Available -> value
        is Initial -> value
        is Replacing -> lastKnown
        is Unprovided -> lastKnown
    }
}

File: ios/runtime/BridgeValue.swift
// BridgeValue.swift
// Availability wrapper for BridgeKit state values and consume proxies.

/// Availability wrapper for BridgeKit state values and consume proxies.
///
/// - ``available(_:)``: A live provider is present; value comes from the provider.
/// - ``initial(_:)``: No provider yet (or not yet connected); value is the
///   DSL-declared initial. Used before the first provider connects.
/// - ``replacing(_:)``: Provider is reconnecting; last known value is still
///   accessible but marked stale. Transitions to ``available`` if re-provision
///   arrives, or to ``unprovided`` after the grace window expires.
/// - ``unprovided(_:)``: Provider existed but was closed/disconnected; `lastKnown`
///   carries the last observed value for graceful degradation.
public enum BridgeValue<T> {
    /// A live provider is present.
    case available(T)
    /// No provider yet; value is the DSL-declared initial.
    case initial(T)
    /// Provider is reconnecting; `lastKnown` is the last value from the closing provider.
    case replacing(T?)
    /// Provider was closed/disconnected; `lastKnown` is the last observed value.
    case unprovided(T?)

    /// Unwrap the value regardless of availability status.
    /// Returns `nil` for ``replacing(_:)`` / ``unprovided(_:)`` when no prior value exists.
    public func valueOrNil() -> T? {
        switch self {
        case .available(let value):   return value
        case .initial(let value):     return value
        case .replacing(let last):    return last
        case .unprovided(let last):   return last
        }
    }

    /// Re-type the carried value via `transform`, preserving availability state.
    ///
    /// Used by generated outbound State getters to convert the runtime's
    /// `BridgeValue<Any?>` stream into the contract's concrete `BridgeValue<U>`
    /// without an invariant `as!` cast on the whole `AsyncStream`.
    ///
    /// `.available`/`.initial` degrade to `.unprovided(nil)` when `transform`
    /// returns `nil` (decode failure), so a malformed value never traps.
    public func remap<U>(_ transform: (T) -> U?) -> BridgeValue<U> {
        switch self {
        case .available(let value):   return transform(value).map(BridgeValue<U>.available) ?? .unprovided(nil)
        case .initial(let value):     return transform(value).map(BridgeValue<U>.initial) ?? .unprovided(nil)
        case .replacing(let last):    return .replacing(last.flatMap(transform))
        case .unprovided(let last):   return .unprovided(last.flatMap(transform))
        }
    }
}
